### ISSUE TITLE
Empty string literals in Marshal.cs

### ISSUE THREAD

**Issue by tudoranghelina91** [dotnet/runtime#68347]
While working on #34056 I noticed that for some exceptions inside `System.Runtime.InteropServices.Marshal` some exceptions get thrown with empty string literals. Wouldn't it be better for these empty string literals to be replaced with the string.Empty constant?

e.g.
![image](https://user-images.githubusercontent.com/16164903/164531111-677293e1-4060-4515-a57f-403605e2d013.png)

**Comment by jkotas** [comment]
There is no practical functional or performance difference between `string.Empty` and `""` (generates negligibly smaller IL). It is a matter of personal preference on which one to use.

For reference, the libraries subdirectory in this repo has 6_000 instances of `string.Empty` and 21_000 instances of `""`.

**Comment by danmoseley** [comment]
these can be replaced with `throw new ObjectDisposedException()` though.

**Comment by jkotas** [comment]
> throw new ObjectDisposedException() though.

`ObjectDisposedException` does not have a parameter-less constructor.

**Comment by jkotas** [comment]
It can be replaced with `new ObjectDisposedException(null)` that's tiny bit better than empty string.

**Comment by danmoseley** [comment]
Oh - I missed that it's private. Yup.

**Comment by tudoranghelina91** [comment]
> It can be replaced with `new ObjectDisposedException(null)` that's tiny bit better than empty string.

Well, it makes more sense to me as well. I'm up to take this one too since I brought it up

**Comment by jkotas** [comment]
Feel free to include this change in the PR you are working on.

**Comment by jyller** [comment]
> There is no practical functional or performance difference between `string.Empty` and `""` (`""` generates negligibly smaller IL). It is a matter of personal preference on which one to use.
> 
> For reference, the libraries subdirectory in this repo has 6_000 instances of `string.Empty` and 21_000 instances of `""`.

What am I supposed to make of this sentence in brackets, "" generates negligibly smaller IL

**Comment by jkotas** [comment]
First use of `string.Empty` in an assembly adds about 30 bytes to the compiled binary. It is cost of encoding field reference. `""` does not have this cost. Subsequent uses of `string.Empty` have same binary size cost as `""`.

The sections in .NET binaries are padded to multiples of 512 bytes. It means that these extra 30 bytes are observable only if they push the binary into having a new 512 chunk.

**Comment by danmoseley** [comment]
Would it make sense to bulk replace to "" purely for consistency sake?

**Comment by jkotas** [comment]
> As I can see from the code below, using String.Empty twice, S2 and S3 have the same IL overhead. Am I misinterpreting this?

The IL text does not show you the binary size.

The experiment you can run is:

1. Compile project that has just a single type and single method that returns `""`.
2. Look at the size of `.text` sections using a tool like `dumpbin`. `dumpbin` it is part of Visual Studio C++ workload. Run `dumpbin /HEADERS <your test name>.dll` and look for:
```
...
   .text name
     934 virtual size
...
````
3. Change the method to return `string.Empty` and compile the project again
4. Look at the size of `.text` sections in the produced .dll again. You should see that it is about 30 bytes larger.

**Comment by jyller** [comment]
I created a new class library
```c#
using System;

namespace ClassLibrary1
{
    class C
    {
        string M()
        {
            return string.Empty;
        }
    }
}
```

```
.\dumpbin.exe /HEADERS E:\ConsoleApp3\ClassLibrary1\bin\Debug\netcoreapp3.1\ClassLibrary1.dll
```

use ""
```
SECTION HEADER #1
   .text name
     644 virtual size
```
use string.Empty
```
SECTION HEADER #1
   .text name
     660 virtual size
```

Why is there 16 bytes difference here.

**Comment by jyller** [comment]
> First use of `string.Empty` in an assembly adds about 30 bytes to the compiled binary. It is cost of encoding field reference. `""` does not have this cost. Subsequent uses of `string.Empty` have same binary size cost as `""`.
> 
> Sections in .NET binaries are padded to multiples of 512 bytes. It means that these extra 30 bytes are observable only if they push the binary into having a new 512 chunk.

Thank you very much for answering my question. I would like to know the same cost as "" in the follow-up. How do I understand this, and why is it 30 instead of others.

**Comment by jkotas** [comment]
To understand the detailed reason why this difference exists, you would need to drill into the details of the binary encodings in https://www.ecma-international.org/publications-and-standards/standards/ecma-335/

The exact observed cost of the `string.Empty` reference (16 bytes you have seen vs. 30 bytes that I have quoted) varies due to number of subtle factors. For example, there are internal paddings used within the .text section, and so the exact cost of the `string.Empty` reference may vary based on the names of other types and methods in the binary.

The bottom line is that the cost is negligible, and it is not something to worry about.

**Comment by tudoranghelina91** [comment]
> @tudoranghelina91 do you plan to offer a PR to replace all use of "ObjectDisposedException(XXX)" where XXX is "" or String.Empty, with null?

Hey @danmoseley, I will include these changes with the same issue that has been reffered to (as previously suggested)

**Comment by akoeplinger** [comment]
@tudoranghelina91 do you still plan to work on this or should we move it back to "help wanted"?

**Comment by tudoranghelina91** [comment]
> @tudoranghelina91 do you still plan to work on this or should we move it back to "help wanted"?

I thought that this was taken care of in the issue that I was working on in which I discovered this. But if it hasn't, I can start working on it. Please let me know so I can start working on it
